[Feedly] Initial Integration

Mathieu4141 opened the pull request:
## Contributing to Cortex XSOAR Content

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
N/A

## Description
Adding an integration with Feedly. The integration will pull articles with entities, IoCs and their relationships, and create the corresponding indicators in XSOAR.

## Must have
- [x] Tests
- [x] Documentation

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/demisto/content?pullRequest=29822) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/demisto/content?pullRequest=29822) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/demisto/content?pullRequest=29822) it.</sub>

MichaelYochpaz commented:
Hi @Mathieu4141, thank you for your contribution.
Your PR review will be slightly delayed since it is the holiday season in Israel, and we have a holiday this weekend (21th-25th of September), I apologize in advance.

MichaelYochpaz commented:
Hey @Mathieu4141, thank you for the contribution.
Please fill the [contribution form](https://github.com/demisto/content/pull/29822#issuecomment-1729739901). Without it, we cannot start the contribution process.

Mathieu4141 commented:
> Hey @Mathieu4141, thank you for the contribution. Please fill the [contribution form](https://github.com/demisto/content/pull/29822#issuecomment-1729739901). Without it, we cannot start the contribution process.

Done! Sorry I was missing some details to fill it before 
Let me know if you need anything more before reviewing

MichaelYochpaz commented:
@Mathieu4141 Hey, looks like it's good to go for a review now, and you've completed all the required tasks :)
Please note that there might be a few days delay in the review process since the contribution PRs queue is a bit overloaded.
I apologize in advance

MichaelYochpaz commented on Packs/Feedly/Integrations/Feedly/test_data/FeedHelloWorld_mock.txt:
Is this file specific to your feed, or one left over from the example `HelloWorld` feed?
If it's used by your unit-tests, please rename it to something more meaningful. If not - remove it.

MichaelYochpaz reviewed (commented):
Hey, I'm still reviewing the code, but there are a few notes I made that can be addressed in the meantime.

Please also pull and merge from master to your branch, and fix the errors here:
https://github.com/demisto/content/actions/runs/6263315560/job/17011882198?pr=29822

Feel free to reach out to me here or on the DFIR Slack if you need any help.

Mathieu4141 commented on Packs/Feedly/.secrets-ignore:
```diff
@@ -0,0 +1,3 @@
+https://api.feedly.com
+https://feedly.com
+base_url=FEEDLY_BASE_URL,
```
yes it was detected as a secret

MichaelYochpaz reviewed (commented):
Some additional notes

MichaelYochpaz commented on Packs/FeedFeedly/Integrations/FeedFeedly/FeedFeedly.py:
```diff
@@ -0,0 +1,1013 @@
+    )
+
+
+def main():
```
To pass the coverage requirement (unit-tests aren't relevant for the main funciton)
```suggestion
def main():  # pragma: no cover
```

MichaelYochpaz commented on Packs/FeedFeedly/Integrations/FeedFeedly/FeedFeedly.py:
```diff
@@ -0,0 +1,1013 @@
+    return client.fetch_indicators_from_stream(
+        params["feedly_stream_id"],
+        newer_than=float(context.get("last_successful_run", time.time() - 7 * 24 * 3600)),
+        # newer_than=time.time() - 7 * 24 * 3600,
```
Was this left here by mistake? If so:
```suggestion
```

MichaelYochpaz reviewed (commented):
Some additional notes following the last changes. (seems like you've replaced the name of the files using the IDE, and it changed them in some unwanted places?)

Please also note the [failing validations](https://app.circleci.com/pipelines/github/demisto/content/307218/workflows/122090d1-7827-4696-bd6e-337d185ba02d/jobs/722595) and [linting error](https://app.circleci.com/pipelines/github/demisto/content/307218/workflows/122090d1-7827-4696-bd6e-337d185ba02d/jobs/722598).

MichaelYochpaz commented:
@Mathieu4141 Hey, you have [2 failing validations](https://app.circleci.com/pipelines/github/demisto/content/307733/workflows/933d955d-eac4-480a-990c-92be0adb29c1/jobs/723064):
1. A mypy error:
```
/home/circleci/project/Packs/FeedFeedly/Integrations/FeedFeedly/FeedFeedly.py:691:72: error:
Argument 1 to "is_valid_type" of "RelationshipsTypes" has incompatible type
"Optional[Any]"; expected "str"  [arg-type]
    ...ntityRelationship.RelationshipsTypes.is_valid_type(relationship_type):
                                                          ^~~~~~~~~~~~~~~~~
Found 1 error in 1 file (checked 1 source file)
```
2. Your unit-test coverage is too low. We require a minimum of 70% coverage (currently at 59.8%).

Please fix :)

MichaelYochpaz reviewed (commented):
Hey @Mathieu4141, the demo went well, good job!

Notes following the demo:
- Move the `Trust any certificate (not secure)` and `Use system proxy settings` parameters to be the bottom of the configuration page (their order in the UI is the same as the YAML).
- When using the `limit` parameter, it seems like it's always missing one (for example, for a limit of 10, it returns 9 values). Please check why that happens and fix it.
- Add the attack pattern IDs and threat actors as tags for the created indicators.
- This one is up to you (just a suggestion, you're not required to change it) - customers are used to the `description` field being a brief description of the indicator, so it's somewhat unfitting to have the full article within that field. One possible solution is to create a custom indicator field that will contain the full article, and a custom layout to decide where it will appear within the UI.

Once these notes are addressed (and the build passes and there are no new issues introduced), PR should be good for a merge :)
Feel free to reach out to me on Slack / email if you have any questions / need help.

Mathieu4141 commented:
@MichaelYochpaz  I addressed all the issues mentionned, and added a new FeedlyReport type with custom layout, so that there is more space for the description of the report

This is how it looks in XSOAR now:
<img width="2124" alt="Screenshot 2023-10-12 at 2 10 42 PM" src="https://github.com/demisto/content/assets/23531109/540b0608-fb31-42b8-8f23-e978e4445843">

Ni-Knight commented:
Hi @Mathieu4141 I would put it a bit different
![image](https://github.com/demisto/content/assets/24267931/2bd4cc17-f564-4483-9300-d518e550050d)

MichaelYochpaz commented:
@Mathieu4141 Also note that I don't see you've addressed the previously mentioned `limit` parameter issue in your recent changes, so I'm assuming it is still unfixed?

Adding @melamedbn to the review since files that are reviewed by our security team (custom layout and indicator type) were added.

Mathieu4141 commented on Packs/FeedFeedly/Integrations/FeedFeedly/FeedFeedly.py:
```diff
@@ -0,0 +1,994 @@
+
+        indicators = STIX2Parser().parse_stix2_objects(objects)
+
+        if limit:
+            indicators = indicators[:limit]
```
@MichaelYochpaz the issue with the limit was that I was applying it on the stix `objects` instead of the XSOAR `indicators`

I think 2 of my intended commits got merged because of unsuccesful pre-commit validation, which is why the fix doesn't appear in my commit messages :)

MichaelYochpaz commented:
@Mathieu4141 Please fix the [following failing validations](https://app.circleci.com/pipelines/github/demisto/content/309034/workflows/00e4d7b0-5da6-4b1e-9b55-b9e6caff7185/jobs/723804).
All should be easily fixed by running:
```
demisto-sdk format -i "Packs/FeedFeedly/Layouts/layoutscontainer-FeedlyReport.json" 
```

Other than that, it looks good to me and ready for a merge (assuming the others don't have any additional notes).
Good job :)

Mathieu4141 commented:
> Looks good to me, good job :) Please make sure to re-test the layout & reports and that they work properly following your last changes.


Yes everything works correctly!

Thanks for all your reviews!

Ni-Knight commented on Packs/FeedFeedly/pack_metadata.json:
```diff
@@ -0,0 +1,22 @@
+        "Data Enrichment & Threat Intelligence"
+    ],
+    "tags": [
+        "Threat Intelligence Management"
```
@Mathieu4141 if this feature is available to non paying customers please add "Free feed" as a tag and add the pack in the "FreeFeed" Pack metadata as well (https://github.com/demisto/content/blob/master/Packs/FreeFeeds/pack_metadata.json)

If not, ignore this message.

Ni-Knight reviewed (commented):
Only one comment about tags.

Ni-Knight reviewed (approved):
Well done.

melamedbn commented:
Hi @Mathieu4141,

We'll be ready if you can fix the panels' positioning in the layout grid, so we won't have blank spaces and add a layout description.
Take the following screenshot as an optional example:

![image](https://github.com/demisto/content/assets/28757135/775549ef-11ca-4605-80a4-e51ac68a52a2)

Adjusted
![image](https://github.com/demisto/content/assets/28757135/2e18df58-2462-4657-af52-e83f2405d9b2)


Best regards,
Ben
